### PR TITLE
Add Hotkey Support

### PR DESCRIPTION
**Pull request by 772807886**
#706 
Add hotkey support, there is a menu item in shadowsocks' menu to set up hotkeys.
Now it can switch enable status, mode, LAN Listener, switch server, and open the log window.

### CODE REVIEW

**Comment by wongsyrone** [comment]
Please write commit title and description in English.

**Comment by wongsyrone** [comment]
Also, write comments in English

File: shadowsocks-csharp/View/HotkeySetting.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                if(e.Shift) {
+                    sb.Append("Shift + ");
+                }
+                if((e.KeyValue >= 33 && e.KeyValue <= 40) ||
+                    (e.KeyValue >= 65 && e.KeyValue <= 90) ||
+                    (e.KeyValue >= 112 && e.KeyValue <= 123)) {
+                    sb.Append(e.KeyCode);
+                } else if(e.KeyValue >= 48 && e.KeyValue <= 57) {
+                    sb.Append('D').Append((char)e.KeyValue);
+                } else if(e.KeyValue >= 96 && e.KeyValue <= 105) {
```
Comment (breakwa11) [inline shadowsocks-csharp/View/HotkeySetting.cs]: what these numbers means?

Comment (772807886) [inline shadowsocks-csharp/View/HotkeySetting.cs]: It mean the key code.
33~40 is the function keys on keyboard.
65~90 means A to Z on keyboard.
112~123 means F1 to F12 on keyboard.
48~57 means 0 to 9 number keys above letter keys on keyboard.
96~105 means 0 to 9 number keys at right of the keyboard.
You can google "Keycode" to get more information.

**Comment by 772807886** [comment]
I rebased the commits with commit title and description in English.

File: shadowsocks-csharp/View/HotkeySetting.cs (hunk 1 above)
Comment (breakwa11) [inline shadowsocks-csharp/View/HotkeySetting.cs]: I means you should use constant instead

Comment (772807886) [inline shadowsocks-csharp/View/HotkeySetting.cs]: OK, I will change it.

File: shadowsocks-csharp/Controller/HotKey.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Windows.Forms;
+
+namespace Shadowsocks.Controller {
+    public class HotKey : Form {
+        /// <summary>
+        /// 接受热键的窗体
```
Comment (wongsyrone) [inline shadowsocks-csharp/Controller/HotKey.cs]: These code comments should also in English.

**Comment by wongsyrone** [comment]
I think about your implementation, finally, come up with another idea.

We don't have to handle Win API ourselves, https://github.com/kirmir/GlobalHotKey is handy after reading his code, it is in NuGet packages. Shadowsocks is a tray application; it is hard to maintain Form's handle, although creating a hidden form is possible.

You can rewrite your code based on https://github.com/wongsyrone/shadowsocks-windows/commits/shortcut, or in other ways follow the idea. I want this feature behaves like QQ's setting hotkey form.

Any idea?

**Comment by 772807886** [comment]
Ok, I merged your branch into my pull request.

**Comment by wongsyrone** [comment]
I pushed a commit to finalize hotkey format, you can check commit message and the test cases.

**Comment by wongsyrone** [comment]
Did you delete all your changes when rebasing?

**Comment by 772807886** [comment]
I had backup the folder before I rebase.
After rebase, all of changes was deleted.

**Comment by wongsyrone** [comment]
Close due to conflicts and inactive.
You can check https://github.com/shadowsocks/shadowsocks-windows/pull/724.

I changed the server switching logic. Yours won't work if the user configured more than ten servers. I change this to move to prev/next server. This is a workaround to satisfy users who add many servers for backup.

Also, a user should register all hotkeys explicitly when starting Shadowsocks, since environment may vary. If you want to open multiple instances of Shadowsocks, you have to change hotkeys for another instance. Otherwise, it will not work.

**Pull request closed without merging**